File: Java/day22.java
import java.util.*;
import java.io.*;

class Node {
    Node left, right;
    int data;
    Node(int data) {
        this.data = data;
        left = right = null;
    }
}

public class day22
{
    public static int getHeight(Node root) {
        // Write your code here
        if (root == null)
            return 0;
        else {
            int left = getHeight(root.left);
            int right = getHeight(root.right);
            return (1 + (left > right ? left : right));
        }
    }

    public static Node insert(Node root, int data){
        if (root == null){
            return new Node(data);
        }
        else {
            Node cur = new Node(data);
            if (data <= root.data){
                cur = insert(root.left, data);
                root.left = cur;
            }
            else{
                cur = insert(root.right, data);
                root.right = cur;
            }
            return root;
        }
    }

    public static void main(String args[]) {
        Scanner sc = new Scanner(System.in);
        int T = sc.nextInt();
        Node root = null;
        while (T-- > 0) {
            int data = sc.nextInt();
            root = insert(root, data);
        }
        int height = getHeight(root);
        System.out.println(height);
    }	
}

File: Scala/day22.scala
// This code does not currently work correctly; could not correctly translate
// the insert() method to Scala's functional paradigm

// Working solutions implemented in Python and Java, see day22.py & day22.java

class Node(d: Int) {
  var left : Node = null
  var right : Node = null
  var data : Int = d
}

object Solution {
  def insert(root: Node, d: Int): Node = {
    if (root == null) {
      val new_node: Node = new Node(d)
      return new_node
    }
    else {
      var root_copy: Node = new Node(root.data)
      var curr: Node = new Node(d)
      if (d <= root_copy.data) {
        curr = insert(root_copy.left, d)
        root_copy.left = curr
      }
      else {
        curr = insert(root_copy.right, d)
        root_copy.right = curr
      }
      return root_copy
    }
  }

  def getHeight(root: Node): Int = {
    // Write your code here
    if (root == null) 0
    else {
      val left: Int = getHeight(root.left)
      val right: Int = getHeight(root.right)
      val max = if (left > right) left else right
      return (1 + max)
    }
  }

  def main(args: Array[String]) {
    val sc = new java.util.Scanner(System.in)
    val T: Int = sc.nextInt()
    var root: Node = null
    for (i <- 1 to T) {
      var data = sc.nextInt()
      root = insert(root, data)
    }
    val height: Int = getHeight(root)
    println(height)
  }
}
